joblib/_dask.py

```diff
@@ -163,9 +166,28 @@ def __init__(self, scheduler_host=None, scatter=None,
         else:
             self._scatter = []
             self.data_futures = {}
-        self.task_futures = set()
         self.wait_for_workers_timeout = wait_for_workers_timeout
         self.submit_kwargs = submit_kwargs
+        self.waiting_futures = as_completed(
+            [],
+            loop=client.loop,
+            with_results=True,
+            raise_errors=False
+        )
+        self._results = {}
+        self._callbacks = {}
+
+    async def _collect(self):
+        while self._continue:
+            async for future, result in self.waiting_futures:
+                cf_future = self._results.pop(future)
+                if future.status == "error":
+                    typ, exc, tb = result
+                    cf_future.set_exception(exc)
+                else:
+                    cf_future.set_result(result)
+                self._callbacks.pop(future)(result)
```

Comment:
Please correct me if I'm wrong, but the `joblib` callback contains  a `dask` `client.scatter` call, which will be automatically made into a coroutine if running in the `IOLoop` thread (such as here, as `dask` would detect this call is done in an thread that should not be blocked). Would we want the `callback` to be ran in the `IOLoop`, we would need to make the whole joblib `callback` function execution chain `async` right?

Would it be possible/make sense to run such a `_collect` routine in a regular `thread` instead? This way the callback can remain blocking without blocking the `IOLoop` itself.

Reply:
Thanks for bringing this up.  I hadn't thought of it before.

Currently I see scatter calls both in the constructor, and in the `apply_async` call. My guess is that both of these will be called from the user's main thread, and not from the IOLoop.  If that's the case then I think that we should be safe.  Dask generally runs the IOLoop in a separate thread.

You mention the `callback` function and I have to admit that I don't know what happens in that function.  Does that call Joblib code which might then call `apply_async`?  If so then yes, we might want to queue up the callbacks to run in the main thread (or some other thread)

Reply:
Thanks for the fast feedback.

> You mention the callback function and I have to admit that I don't know what happens in that function.

This callback is actually in charge of dispatching a new task from the `iterator` given as input to the `Parallel` object. Thus, this callback calls `apply_async` which itself triggers a `scatter` call.

Reply:
Ok, we're now always scattering from the event loop, so this should be resolved. 

Oddly, we now sometimes scatter too frequently (maybe because we're now allowing some concurrency) and so there is some inefficiency here that was introduced (see the modified test).  I tried avoiding this with an `asyncio.Lock` but that stopped things.  I'm not yet sure why.

Reply:
Indeed, now the whole callback is done in the event loop. This callback runs `scatter` calls, which are made async (great!), thus non-blocking, but this callback runs also a lot of "not async" joblib code, and I'm afraid we affect performance by running so much joblib code in the event loop (eventhough I'm not an `async` expert).

I coded an alternative here https://github.com/pierreglaser/joblib/commit/00158c8c94bbc9308e4a387c45b2687b61739812, where we should run much less `joblib` code inside the `event` loop. Not saying this is the way to go though.

Reply:
Hrm, alternatively could we call just the callback in another thread?  For debugging reasons it's useful to keep Dask code running the event loop.  Things tend to be smoother generally

How much joblib work happens in the callback?  What are these calls doing?

Reply:
These calls are basically consuming new tasks from the original iterator sent as an input of `Parallel.__call__`. So there's a fair amount of joblib code that this callbacks runs, including code inside `with lock` context managers, scattering the input, and eventually calling `client.submit`. So there is a bunch of `dask` code, and a bunch of `joblib` code run by the callbacks.

> Hrm, alternatively could we call just the callback in another thread?

Doesn't this contradict the fact that we want to keep `dask` code in the event loop since the `callback` contains `dask` (`client.submit`, `client.scatter`) code? 

 See [here](https://github.com/joblib/joblib/blob/dd4b9a0b9b05c137fa089fde41390c0a95877787/joblib/parallel.py#L325-L334).

Reply:
@ogrisel do you have some thoughts on this?

Reply:
I think that for 90% of our users, the task generating iterator will be fast to consume, so this should not be a problem for the majority of our user.

But the iterator would be a lazy data loader to a database or a folder with large compressed files, in which case delegating to another thread is probably a good idea.

Maybe it would be worth experimenting with a synthetic, slow task itetor to see what this would mean for the current design of this PR?

This experiment could probably be turned into an integration test both for the task backend and the other thread or process based backends.